diff -Naur plugin.video.netflix/resources/lib/utils.py plugin.video.netflix2/resources/lib/utils.py
--- plugin.video.netflix/resources/lib/utils.py	2018-11-07 11:57:13.678967152 -0800
+++ plugin.video.netflix2/resources/lib/utils.py	2018-11-07 11:59:02.526967111 -0800
@@ -86,7 +86,7 @@
     if system == 'Windows':
         return base.replace('%PL%', '(Windows NT 6.1; WOW64)')
     # ARM based Linux
-    if platform.machine().startswith('arm'):
+    if platform.machine().startswith('arm') or platform.machine().startswith('aarch'):
         return base.replace('%PL%', '(X11; CrOS armv7l 7647.78.0)')
     # x86 Linux
     return base.replace('%PL%', '(X11; Linux x86_64)')
