diff -Naur kernel-4.9_old/drivers/usb/gadget/function/f_rndis.c kernel-4.9/drivers/usb/gadget/function/f_rndis.c
--- kernel-4.9_old/drivers/usb/gadget/function/f_rndis.c	2019-07-06 20:22:57.734645336 -0700
+++ kernel-4.9/drivers/usb/gadget/function/f_rndis.c	2019-07-06 20:31:05.142045482 -0700
@@ -1000,7 +1000,8 @@
 
 	mutex_init(&opts->lock);
 	opts->func_inst.free_func_inst = rndis_free_inst;
-	opts->net = gether_setup_name_default("rndis");
+	//opts->net = gether_setup_name_default("rndis");
+	opts->net = gether_setup_default();
 	if (IS_ERR(opts->net)) {
 		struct net_device *net = opts->net;
 		kfree(opts);
