diff -Naur l4t-platform-t210-switch_orig/kernel-dts/tegra210-icosa.dts l4t-platform-t210-switch/kernel-dts/tegra210-icosa.dts
--- l4t-platform-t210-switch_orig/kernel-dts/tegra210-icosa.dts	2019-08-11 21:44:49.293714205 -0000
+++ l4t-platform-t210-switch/kernel-dts/tegra210-icosa.dts	2019-08-11 21:46:32.704753972 -0000
@@ -93,6 +93,13 @@
 		};
 	};
 
+        r2p {
+           compatible = "tegra-reboot2payload";
+           status = "okay";
+           default-payload = "reboot_payload.fw";
+           normal-reboot-action = "payload";
+        };
+
 	/delete-node/ psy_extcon_xudc;
 	/delete-node/ usb_cd@7009f000;
 
